File: android/core/data/src/main/kotlin/com/xpx/vault/data/crypto/BackupKeyManager.kt
package com.xpx.vault.data.crypto

import android.content.Context
import java.security.SecureRandom
import javax.crypto.Mac
import javax.crypto.SecretKey
import javax.crypto.spec.SecretKeySpec

/**
 * 备份密钥管理器：
 * - 通过 Argon2id（回退策略可替换）从 PIN 口令 + 设备盐派生 AES-256 BackupKey。
 * - 计算 keyFingerprint = HMAC-SHA256(backupKey, "aivault.fp.v1")[:16]，用于判断备份密码是否变更。
 *
 * 盐值 Salt 首次安装时随机生成 32 字节，持久化在 [EncryptedSharedPreferences]。
 * 导出的 [KdfParams] 仅包含参数，不包含任何密钥/口令字节，可以安全写入备份包头。
 */
class BackupKeyManager(
    private val context: Context,
) {
    /** Kdf 参数。可随备份包头一同持久化。 */
    data class KdfParams(
        val algorithm: String = ALGO_ARGON2ID,
        val saltHex: String,
        val iterations: Int,
        val memoryKb: Int,
        val parallelism: Int,
    )

    /** 派生结果：密钥 + 指纹 + 所用参数（供写入包头）。 */
    data class BackupKeyMaterial(
        val key: SecretKey,
        val fingerprintHex: String,
        val kdfParams: KdfParams,
    )

    /**
     * 读取或初始化 kdf 参数。salt 随机 32B，按设备自适应降级 memory/iterations。
     * 多次调用返回同一 salt。
     */
    fun getOrCreateKdfParams(): KdfParams {
        val prefs = prefs()
        val existingSalt = prefs.getString(KEY_SALT_HEX, null)
        val saltHex = existingSalt ?: run {
            val bytes = ByteArray(SALT_LENGTH_BYTES).also { SecureRandom().nextBytes(it) }
            val hex = bytes.toHex()
            prefs.edit().putString(KEY_SALT_HEX, hex).apply()
            hex
        }
        val existingMem = prefs.getInt(KEY_MEMORY_KB, 0)
        val existingIter = prefs.getInt(KEY_ITERATIONS, 0)
        val existingPar = prefs.getInt(KEY_PARALLELISM, 0)
        return if (existingMem > 0 && existingIter > 0 && existingPar > 0) {
            KdfParams(
                algorithm = prefs.getString(KEY_ALGORITHM, ALGO_ARGON2ID) ?: ALGO_ARGON2ID,
                saltHex = saltHex,
                iterations = existingIter,
                memoryKb = existingMem,
                parallelism = existingPar,
            )
        } else {
            val (memKb, iterations) = Argon2idKdf.chooseParams(context)
            val params = KdfParams(
                algorithm = ALGO_ARGON2ID,
                saltHex = saltHex,
                iterations = iterations,
                memoryKb = memKb,
                parallelism = Argon2idKdf.DEFAULT_PARALLELISM,
            )
            prefs.edit()
                .putString(KEY_ALGORITHM, params.algorithm)
                .putInt(KEY_MEMORY_KB, params.memoryKb)
                .putInt(KEY_ITERATIONS, params.iterations)
                .putInt(KEY_PARALLELISM, params.parallelism)
                .apply()
            params
        }
    }

    /**
     * 派生 BackupKey。[password] 调用方负责用完后 fill(0)。
     */
    fun deriveKey(password: CharArray, params: KdfParams): BackupKeyMaterial {
        val salt = params.saltHex.hexToBytes()
        val keyBytes = Argon2idKdf.derive(
            password = password,
            salt = salt,
            iterations = params.iterations,
            memoryKb = params.memoryKb,
            parallelism = params.parallelism,
            outLen = Argon2idKdf.DEFAULT_KEY_LENGTH_BYTES,
        )
        val key = SecretKeySpec(keyBytes, "AES")
        // 立即清零本地副本；SecretKeySpec 内部会持有自己的 copy。
        keyBytes.fill(0)
        return BackupKeyMaterial(
            key = key,
            fingerprintHex = fingerprint(key),
            kdfParams = params,
        )
    }

    /** keyFingerprint = HMAC-SHA256(backupKey, "aivault.fp.v1")[:16] hex。 */
    fun fingerprint(key: SecretKey): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(key)
        val full = mac.doFinal(FP_DOMAIN_SEPARATOR.toByteArray(Charsets.UTF_8))
        return full.copyOf(FINGERPRINT_LENGTH_BYTES).toHex()
    }

    // Salt 与 kdf 参数本身会随备份包头明文写出，不属于秘密，故使用普通 SharedPreferences，避免 AndroidX Security Crypto 的 MasterKey API 兼容性问题。
    private fun prefs() = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    companion object {
        const val ALGO_ARGON2ID: String = "Argon2id"
        private const val PREFS_NAME = "backup_kdf_prefs"
        private const val KEY_SALT_HEX = "salt_hex"
        private const val KEY_ALGORITHM = "algorithm"
        private const val KEY_MEMORY_KB = "memory_kb"
        private const val KEY_ITERATIONS = "iterations"
        private const val KEY_PARALLELISM = "parallelism"
        private const val SALT_LENGTH_BYTES = 32
        private const val FINGERPRINT_LENGTH_BYTES = 16
        private const val FP_DOMAIN_SEPARATOR = "aivault.fp.v1"

        fun ByteArray.toHex(): String = joinToString("") { b -> "%02x".format(b) }

        fun String.hexToBytes(): ByteArray {
            require(length % 2 == 0) { "hex string length must be even" }
            val out = ByteArray(length / 2)
            for (i in out.indices) {
                out[i] = substring(i * 2, i * 2 + 2).toInt(16).toByte()
            }
            return out
        }
    }
}

File: ios/LumaNox/Core/Crypto/BackupKeyManager.swift
import CryptoKit
import Foundation

/// Backup key derivation — mirrors Android [BackupKeyManager].
final class BackupKeyManager {
    struct KdfParams: Codable, Equatable {
        var algorithm: String
        var saltHex: String
        var iterations: Int
        var memoryKb: Int
        var parallelism: Int

        static let argon2id = "Argon2id"
    }

    struct BackupKeyMaterial {
        let key: Data
        let fingerprintHex: String
        let kdfParams: KdfParams
    }

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    func getOrCreateKdfParams() -> KdfParams {
        if let saltHex = defaults.string(forKey: Keys.saltHex) {
            let mem = defaults.integer(forKey: Keys.memoryKb)
            let iter = defaults.integer(forKey: Keys.iterations)
            let par = defaults.integer(forKey: Keys.parallelism)
            if mem > 0, iter > 0, par > 0 {
                return KdfParams(
                    algorithm: defaults.string(forKey: Keys.algorithm) ?? KdfParams.argon2id,
                    saltHex: saltHex,
                    iterations: iter,
                    memoryKb: mem,
                    parallelism: par
                )
            }
        }
        var salt = Data(count: 32)
        _ = salt.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, 32, $0.baseAddress!) }
        let saltHex = salt.hexString
        let (memKb, iterations) = Argon2idKdf.chooseParams()
        let params = KdfParams(
            algorithm: KdfParams.argon2id,
            saltHex: saltHex,
            iterations: iterations,
            memoryKb: memKb,
            parallelism: Argon2idKdf.defaultParallelism
        )
        defaults.set(params.algorithm, forKey: Keys.algorithm)
        defaults.set(params.saltHex, forKey: Keys.saltHex)
        defaults.set(params.memoryKb, forKey: Keys.memoryKb)
        defaults.set(params.iterations, forKey: Keys.iterations)
        defaults.set(params.parallelism, forKey: Keys.parallelism)
        return params
    }

    func deriveKey(password: String, params: KdfParams) throws -> BackupKeyMaterial {
        let salt = Data(hexString: params.saltHex)
        let key = try Argon2idKdf.derive(
            password: password,
            salt: salt,
            iterations: params.iterations,
            memoryKb: params.memoryKb,
            parallelism: params.parallelism
        )
        return BackupKeyMaterial(
            key: key,
            fingerprintHex: fingerprint(key: key),
            kdfParams: params
        )
    }

    func fingerprint(key: Data) -> String {
        let symmetric = SymmetricKey(data: key)
        let mac = HMAC<SHA256>.authenticationCode(
            for: Data("aivault.fp.v1".utf8),
            using: symmetric
        )
        return Data(mac.prefix(16)).hexString
    }

    private enum Keys {
        static let saltHex = "backup_kdf_salt_hex"
        static let algorithm = "backup_kdf_algorithm"
        static let memoryKb = "backup_kdf_memory_kb"
        static let iterations = "backup_kdf_iterations"
        static let parallelism = "backup_kdf_parallelism"
    }
}

extension Data {
    var hexString: String {
        map { String(format: "%02x", $0) }.joined()
    }

    init(hexString: String) {
        precondition(hexString.count % 2 == 0)
        var bytes = [UInt8]()
        bytes.reserveCapacity(hexString.count / 2)
        var index = hexString.startIndex
        while index < hexString.endIndex {
            let next = hexString.index(index, offsetBy: 2)
            let byte = UInt8(hexString[index..<next], radix: 16) ?? 0
            bytes.append(byte)
            index = next
        }
        self = Data(bytes)
    }
}
